### PR TITLE
pure-maps: fix

### PR DESCRIPTION
**Pull request by dotlambda**
###### Motivation for this change
closes https://github.com/NixOS/nixpkgs/pull/116209

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 117045` at ebd9adf9 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built successfully:</summary>
  <ul>
    <li><a href="https://gist.github.com/fc575e798dc0e292828983c7dd5c6169">pure-maps</a></li></li>
    <li><a href="https://gist.github.com/6c7ff91fb194923173c27a261a0aae93">python38Packages.geomag</a></li></li>
    <li><a href="https://gist.github.com/e5e9090ce8a0de8f4c88f695cf12e068">python39Packages.geomag</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/applications/misc/pure-maps/default.nix:19:5:
```
   |
19 |     ./unvendor.patch
   |     ^
```
</li>
  </ul>
</details>

File: pkgs/development/python-modules/geomag/default.nix [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  version = "2020-05-27";
+
+  src = fetchFromGitHub {
+    owner = "rinigus";
+    repo = "geomag";
+    rev = "8eb9a730c8643fb7d63fdee4fd9a195ee8ba4df2";
+    sha256 = "1c1s75g3xfjib396y486sdbqsp4rkpvvcla3j2221symdx1l29s2";
+  };
+
+  propagatedBuildInputs = lib.optional (pythonOlder "3.2") functools32;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/geomag/default.nix:20]:
```suggestion
  propagatedBuildInputs = lib.optionals (pythonOlder "3.2") [ functools32 ];
```

FYI for future PRs: this is preferred because it is more strict and less likely to do mistakes with.

Comment (dotlambda) [reply pkgs/development/python-modules/geomag/default.nix:20]: What's more strict about it? I don't see any advantage.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 117045` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
3 packages built:
</summary>
<ul>
<li>pure-maps</li>
<li>python38Packages.geomag</li>
<li>python39Packages.geomag</li>
</ul>
</details>

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -2640,6 +2640,8 @@ in {
 
   geojson-client = callPackage ../development/python-modules/geojson-client { };
 
+  geomag = callPackage ../development/python-modules/geomag { };
```
Comment (Thra11) [inline pkgs/top-level/python-packages.nix:2643]: You appear to be putting a fork (albeit a more actively-developed one) into the global python package namespace. This could be confusing for anyone who expects geomag to be the upstream version which appears in pypi, or awkward if somebody wants to package something in the future which depends on upstream geomag.

Not saying we _can't_ have rinigus' geomag as pythonPackages.geomag, but it's worth thinking through carefully.

File: pkgs/development/python-modules/geomag/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/geomag/default.nix:20]: optionals makes sure that the 2nd argument is a list. optional could be anything.

Comment (Thra11) [reply pkgs/development/python-modules/geomag/default.nix:20]: Pedantically speaking, they both return a list (either empty or containing the second argument), and I don't think either cares what the types of the list contents are, so there is no meaningful difference in terms of types: `lib.optional true garbage` returns `[ garbage ]`, while `lib.optionals true [ garbage ]` also returns `[ garbage ]`.

File: pkgs/top-level/python-packages.nix (hunk 2 above)
Comment (dotlambda) [reply pkgs/top-level/python-packages.nix:2643]: Maybe we should encourage rinigus to publish it on PyPI.
It seems like the data bundled with the package is outdated in older versions and that's the only thing this fork updates. Hence it doesn't seem problematic.

Comment (Thra11) [reply pkgs/top-level/python-packages.nix:2643]: Yeah. The original 'upstream' hasn't been updated since 2014, so I'm inclined to consider it abandoned / unsupported, with the fork and the fork fork appearing to be maintained and updated versions of it, and thus a reasonable candidate to package as simply 'geomag' (In situations where upstream is still going and the fork has just decided to go in a different direction, I would be more opposed to the fork being the main package in nixpkgs).

Comment (dotlambda) [reply pkgs/top-level/python-packages.nix:2643]: @rinigus How do you feel about publishing your fork of geomag on PyPI?

Comment (rinigus) [reply pkgs/top-level/python-packages.nix:2643]: Not thrilled about it. As Pure Maps has been rewritten using C++ main(), I may switch to some other library for that info. Not the highest priority, but would prefer that than to start packaging someones outdated code which probably has functionality that I don't use. And, as a result, may have bugs that I am not hitting.

Comment (dotlambda) [reply pkgs/top-level/python-packages.nix:2643]: Understandable. Thanks for your great work anyway!

**Comment by Thra11** [comment]
Can the first commit stand on its own, and if so, would you consider splitting this into two separate PRs? I'm not sure I agree with splitting out a vendored dependency just to essentially manually re-vendor it ourselves. By that I mean that:
* The added pythonPackages.geomag is not intended for general use: the developer maintains it solely as a dependency of pure-maps, and it would be rude of us to expect them to do any more than that (or to expose it in such a way that our users have such expectations)
* Since nothing else depends on it, and it's maintained by the same developer as pure-maps, it will effectively be tied to pure-maps and is extremely unlikely to be updated to any version other than what the latest version of pure-maps specifies. This means that most of the benefits of de-vendoring a dependency don't apply. For example, if there were a vulnerability discovered and fixed in geomag, pure-maps's vendored geomag will get the fix just as fast as our de-vendored version would.
* Essentially, the way I see it, geomag may be based on code developed elsewhere, but for all practical purposes it's now just part of the pure-maps codebase that happens to be stored as a git submodule.

**Comment by dotlambda** [comment]
Okay, agreed.

**Review by Thra11 — APPROVED** [review]
Looks good to me!
